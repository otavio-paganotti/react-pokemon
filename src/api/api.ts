import axios, { AxiosInstance } from 'axios';

export const http: AxiosInstance = axios.create({
  baseURL: String(import.meta.env.VITE_APP_API_URL || ''),
});
