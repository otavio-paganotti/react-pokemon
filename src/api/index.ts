/**
 * Exporting the Axios Instance
 */
export * from './api';

/**
 * Exporting all modules
 */
export * as Pokemon from './services/rest/pokemon';
