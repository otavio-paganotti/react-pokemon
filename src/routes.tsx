import type { RouteObject } from 'react-router';
import MainLayout from '@/layouts/Default';
import Home from '@/pages/index/Index'
import Pokemon from '@/pages/pokemon/Index'

const routes: RouteObject[] = [
  {
    path: '*',
    element: <MainLayout />,
    children: [
      {
        index: true,
        element: <Home />,
      },
      {
        index: false,
        path: ':id',
        element: <Pokemon />
      }
    ],
  },
];

export default routes;