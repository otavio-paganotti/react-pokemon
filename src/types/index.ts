export type Nullable<T = unknown> = T | null;

export interface PokemonItem {
  name: string;
  url: string;
  id: number;
  content: Pokemon;
}

export interface Response<T> {
  count: number;
  next: Nullable<string>;
  previous: Nullable<string>;
  results: T[];
  pagination: {
    firstPage: number;
    lastPage: number;
    nextPage: number;
    prevPage: number;
    nextPages: number[];
    prevPages: number[];
    page: number;
  };
}

export interface Sprites {
  back_default: string;
}

export interface Stat {
  base_stat: number;
  effort: number;
  stat: Species;
}

export interface Type {
  slot: number;
  type: Species;
}

export interface VersionGroupDetail {
  level_learned_at: number;
  move_learn_method: Species;
  version_group: Species;
}

export interface Move {
  move: Species;
  version_group_details: VersionGroupDetail[];
}

export interface VersionDetail {
  rarity: number;
  version: Species;
}

export interface Species {
  name: string;
  url: string;
}

export interface HeldItem {
  item: Species;
  version_details: VersionDetail[];
}

export interface GameIndex {
  game_index: number;
}

export interface Ability {
  ability: Species;
  is_hidden: boolean;
  slot: number;
}

export interface Pokemon {
  abilities: Ability[];
  base_experience: number;
  forms: Species[];
  game_indices: GameIndex[];
  height: number;
  held_items: HeldItem[];
  id: number;
  is_default: boolean;
  location_area_encounters: string;
  moves: Move[];
  name: string;
  order: number;
  past_types: unknown[];
  species: Species;
  sprites: Sprites;
  stats: Stat[];
  types: Type[];
  weight: number;
}

export interface IPagination {
  firstPage: number;
  lastPage: number;
  nextPages: number[];
  prevPages: number[];
  nextPage: number;
  prevPage: number;
  page: number;
}
