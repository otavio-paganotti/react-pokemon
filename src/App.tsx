import routes from '@/routes';
import { useRoutes } from 'react-router-dom';
import '@/assets/css/main.scss';

const App = () => {
  const content = useRoutes(routes);

  return content;
};

export default App;
