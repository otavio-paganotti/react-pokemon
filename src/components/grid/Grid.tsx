import Paginator from './Paginator';
import List from './List';
import { useLoader } from '@/hooks';
import { useEffect, useState } from 'react';
import { PokemonItem, Response } from '@/types';

const GenericGrid = <T extends PokemonItem>({
  module,
}: {
  module: (page?: number) => Promise<Response<T>>;
}) => {
  const [initial] = useState<Response<T>>({
    count: 0,
    next: null,
    previous: null,
    results: [],
    pagination: {
      firstPage: 1,
      lastPage: 1,
      nextPage: 1,
      prevPage: 1,
      nextPages: [1],
      prevPages: [1],
      page: 1,
    },
  });

  const [loader, data, isLoading] = useLoader<Response<T>>(initial);

  useEffect(() => {
    loader(module);

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const page = (page: number) => {
    if (isLoading) return;

    loader(() => module(page));
  };

  const handleSearch = () => {
    loader(() =>
      Promise.resolve({
        count: 0,
        next: null,
        previous: null,
        results: [],
        pagination: {
          firstPage: 0,
          lastPage: 0,
          nextPage: 0,
          prevPage: 0,
          nextPages: [0],
          prevPages: [0],
          page: 0,
        },
      })
    );
  };

  const handleClearSearch = () => {
    loader(module);
  };

  return (
    <div>
      {isLoading && <div>Loading...</div>}
      <div className="p-4 h-screen-minus-56 grid grid-cols-1 content-between gap-4">
        {!isLoading && (
          <>
            <div className="flex gap-4 justify-between">
              <input
                type="search"
                className="w-full px-4 py-2"
                placeholder="Busque o PokÃ©mon..."
              />
              <button
                className="px-4 py-2 text-white"
                onClick={handleSearch}
                style={{ backgroundColor: 'blue' }}
              >
                Buscar
              </button>
              <button
                className="px-4 py-2"
                onClick={handleClearSearch}
                style={{ backgroundColor: 'gray', color: 'black' }}
              >
                Limpar
              </button>
            </div>
            <List items={data?.results} />
          </>
        )}
        <Paginator pagination={data?.pagination} page={page} />
      </div>
    </div>
  );
};

export default GenericGrid;
