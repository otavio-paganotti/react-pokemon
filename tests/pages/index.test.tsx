import React from 'react';
import renderer from 'react-test-renderer';
import IndexComponent from '../../src/pages/index/Index';
import Grid from '../../src/components/grid/Grid';
import { describe, beforeEach, expect, test, vi } from 'vitest';
import { Pokemon } from '../../src/api';

vi.mock('../../src/components/grid/Grid');

const toJson = (component: renderer.ReactTestRenderer) => {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
};

const onMounted = async () => {
  let component;

  component = renderer.create(
    <IndexComponent />
  );

  return toJson(component);
};

describe('Index.tsx', () => {
  let tree;

  beforeEach(async () => {
    vi.clearAllMocks();

    tree = await onMounted();
  });

  test('should render correctly the Index component and take a snapshot', () => {
    expect(tree).toMatchSnapshot();
  });

  test('should have 1 Grid child element must be called', async () => {
    expect(Grid).toHaveBeenCalled();

    expect(Grid).toHaveBeenCalledTimes(1);

    expect(Grid).toHaveBeenCalledWith({
      module: Pokemon.getPokemons,
    }, {});
  });
});
