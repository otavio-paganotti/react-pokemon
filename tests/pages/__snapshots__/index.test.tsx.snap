// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Index.tsx > should render correctly the Index component and take a snapshot 1`] = `
<div
  className="mx-auto max-w-7xl sm:px-6 lg:px-8"
/>
`;
